Use Angular Providers in boostrapApplication option

valentinpalkovic opened the pull request:
Issue: N/A

Use Angular Providers in bootstrapApplication option and rework the way of how imports are initialized for the Angular Storybook standalone wrapper.

<!-- Thank you for contributing to Storybook! If your PR is related to an issue, provide the number(s) above, e.g. #1000, #1001 -->

## What I did

<!-- Briefly describe what your PR does -->

## How to test

<!-- Please include the steps to test your changes here. For example:

1. Run a sandbox for template, e.g. `yarn task --task sandbox --start-from auto --template react-vite/default-ts`
2. Open Storybook in your browser
3. Access X story

-->

## Checklist

<!-- Please check (put an "x" inside the "[ ]") the applicable items below to make sure your PR is ready to be reviewed. -->

- [ ] Make sure your changes are tested (stories and/or unit, integration, or end-to-end tests)
- [ ] Make sure to add/update documentation regarding your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

#### Maintainers

- [ ] If this PR should be tested against many or all sandboxes,
      make sure to add the `ci:merged` or `ci:daily` GH label to it.
- [ ] Make sure this PR contains **one** of the labels below.

`["cleanup", "BREAKING CHANGE", "feature request", "bug", "documentation", "maintenance", "dependencies", "other"]`

<!--

Everybody: Please submit all PRs to the `next` branch unless they are specific to the current release. Storybook maintainers cherry-pick bug and documentation fixes into the `main` branch as part of the release process, so you shouldn't need to worry about this. For additional guidance: https://storybook.js.org/docs/react/contribute/how-to-contribute

-->

valentinpalkovic committed:
fix imports/providers without caching componentNgModule

fix imports/providers without caching componentNgModule

sheriffMoose commented:
@valentinpalkovic I updated the mechanism to not cache the componentNgModule while better process the imports and providers separately

valentinpalkovic commented:
Now, this kind of error is back:

```
Type FormComponent is part of the declarations of 2 modules: StorybookComponentModule and StorybookComponentModule! Please consider moving FormComponent to a higher module that imports StorybookComponentModule and StorybookComponentModule. You can also create a new NgModule that exports and includes FormComponent, then import that NgModule in StorybookComponentModule and StorybookComponentModule.
```

As I told you yesterday, this error is pretty strange because although Standalone components are used and bootstrapped independently from each other, they still somehow infer. Take for example a look at http://localhost:6006/?path=/docs/example-button--docs. Only the first button on the top is rendered. All the other Stories at the bottom aren't

sheriffMoose committed:
add PropertyExtractor.ts for extracting imports/declarations/providers

`imports`/`declarations` will be used only once per story file.
while `providers` will be used for each story rendered.

valentinpalkovic commented on code/frameworks/angular/src/client/angular-beta/utils/PropertyExtractor.ts:
```diff
@@ -0,0 +1,123 @@
+ *
+ * - Executes recursively
+ *
+ * @date `1/24/2023 - 12:58:03 PM`
+ * @author `sheriffMoose`
```
I'm not sure if this was added by your IDE or manually, but usually, for readability reasons, we don't add author information to each function or method. Second, this information has to be updated every time another developer touches the code. Let's remove this.

valentinpalkovic commented on code/frameworks/angular/src/client/angular-beta/utils/PropertyExtractor.ts:
```diff
@@ -0,0 +1,123 @@
+import { isDeclarable, isStandaloneComponent } from './NgComponentAnalyzer';
+import { isComponentAlreadyDeclared } from './NgModulesAnalyzer';
+
+/**
+ * Flattens Array Deeply
+ *
+ * - Executes recursively
```
Also, let's remove comments for self-explanatory and self-explaining functions. In this case, the name of the function is self-explaining. Let's only add documentation if it is indispensable.

sheriffMoose replied:
👍

valentinpalkovic commented on code/frameworks/angular/src/client/angular-beta/utils/PropertyExtractor.ts:
```diff
@@ -0,0 +1,123 @@
+ * @date `1/24/2023 - 12:58:03 PM`
+ * @author `sheriffMoose`
+ *
+ * @param {any[]} arr - Array to be flattened
+ * @returns {*} Array of imports
```
Usually, we don't need this kind of JSDOC annotation; at least the type information is provided by Typescript. Again, let's only add descriptions for params and returns if it is crucial and not self-explaining.

sheriffMoose replied:
Sure

valentinpalkovic commented on code/frameworks/angular/src/client/angular-beta/utils/PropertyExtractor.ts:
```diff
@@ -0,0 +1,123 @@
+import { CommonModule } from '@angular/common';
```
I really like the encapsulation. Well done!

valentinpalkovic commented:
@sheriffMoose Well done  🥳 I Left some comments. Did you figure out the caching issue locally?

sheriffMoose commented:
@valentinpalkovic  yep, I created a new empty clone on a different machine

sheriffMoose commented on code/frameworks/angular/src/client/angular-beta/utils/PropertyExtractor.ts:
```diff
@@ -0,0 +1,123 @@
+ * @param {any[]} arr - Array to be flattened
+ * @returns {*} Array of imports
+ */
+const flattenArray = (arr: any[]) => {
```
Yep, that would be better indeed, I will make that change shortly

sheriffMoose committed:
fixes & updates as per @valentinpalkovic comments

valentinpalkovic commented on code/frameworks/angular/src/client/angular-beta/utils/PropertyExtractor.ts:
```diff
@@ -47,16 +17,13 @@ const uniqueArray = (arr: any[]) => {
- * @author `sheriffMoose`
- *
  * @param {NgModuleMetadata} metadata - NG Module Metadata
  * @returns {*} Array of imports
```
Let's remove this one as well

valentinpalkovic commented on code/frameworks/angular/src/client/angular-beta/utils/PropertyExtractor.ts:
```diff
@@ -76,16 +43,14 @@ export const extractImports = (metadata: NgModuleMetadata) => {
- * @author `sheriffMoose`
- *
  * @param {NgModuleMetadata} metadata - NG Module Metadata
  * @returns {*} Array of providers
```
Let's remove this one as well

valentinpalkovic commented on code/frameworks/angular/src/client/angular-beta/utils/PropertyExtractor.ts:
```diff
@@ -100,9 +65,6 @@ export const extractProviders = (metadata: NgModuleMetadata): Provider[] => {
- *
  * @param {NgModuleMetadata} metadata
  * @param {?*} storyComponent
  * @returns {*} Array of declarations
```
Let's remove this one as well

sheriffMoose committed:
update jsdoc

valentinpalkovic committed:
Disable analyzeSourceFiles in NX
